File: src/engine/CompiledMesh.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace SimpleEngine
{
    public abstract class Poseable
    {
        public abstract CompiledPose CurrentPose
        {
            get;
        }
    }
    public class CompiledMesh : Poseable
    {
        List<MeshPart> parts = new List<MeshPart>();

        public CompiledMesh(MediaLoader ml, MeshDef def)
        {
            List<Exception> errors = new List<Exception>();
            foreach (MeshDef.MaterialDef m in def.Materials)
            {
                if (def.hasTrianglesFor(m))
                {
                    try
                    {
                        parts.Add(new MeshPart(ml, m, def, this));
                    }
                    catch (Exception e)
                    {
                        errors.Add(e);
                    }
                }
            }

            if (errors.Count != 0)
            {
                throw new Exception("Several errors occured");
            }
        }

        public void sendToRenderer(RenderList r, vec3 pos, quat rot)
        {
            foreach (MeshPart part in parts)
            {
                r.add(part, pos, rot);
            }
        }

        CompiledPose pose = null;

        public void setPose(CompiledPose pose)
        {
            this.pose = pose;
        }

        public override CompiledPose CurrentPose
        {
            get { return pose; }
        }
    }
}

File: src/engine/compiledmesh.cc
#include "engine/compiledmesh.h"

#include <memory>
#include <stdexcept>

#include "engine/renderlist.h"

namespace SimpleEngine
{
    CompiledMesh::CompiledMesh(MediaLoader* ml, MeshDef* def)
    {
        std::vector<std::string> errors;
        for (auto& m : def->GetMaterials())
        {
            if (def->HasTrianglesFor(*m))
            {
                try
                {
                    parts.emplace_back(std::make_shared<MeshPart>(ml, *m, def, this));
                }
                catch (const std::runtime_error& e)
                {
                    errors.emplace_back(e.what());
                }
            }
        }

        if (errors.empty() == false)
        {
            throw std::runtime_error("Several errors occured");
        }
    }

    void CompiledMesh::SendToRenderer(RenderList* r, const vec3& pos, const quat& rot)
    {
        for (auto& part : parts)
        {
            r->Add(part, pos, rot);
        }
    }

    void CompiledMesh::SetPose(std::shared_ptr<CompiledPose> new_pose)
    {
        pose = new_pose;
    }

    std::shared_ptr<CompiledPose> CompiledMesh::GetPose()
    {
        return pose;
    }
}
